### PR TITLE
Fix unrestricted reads from keeper

### LINKED ISSUE #60190: Flaky test `02221_system_zookeeper_unrestricted_like`
[Recent fails](https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZEYXkoY2hlY2tfc3RhcnRfdGltZSkgYXMgZCwKY291bnQoKSwgIGdyb3VwVW5pcUFycmF5KHB1bGxfcmVxdWVzdF9udW1iZXIpLCAgYW55KHJlcG9ydF91cmwpCmZyb20gY2hlY2tzIHdoZXJlICcyMDIzLTA4LTI3JyA8PSBjaGVja19zdGFydF90aW1lIGFuZCB0ZXN0X25hbWUgbGlrZSAnJTAyMjIxX3N5c3RlbV96b29rZWVwZXJfdW5yZXN0cmljdGVkX2xpa2UlJyBhbmQgdGVzdF9zdGF0dXMgaW4gKCdGQUlMJywgJ0ZMQUtZJykgZ3JvdXAgYnkgZCBvcmRlciBieSBkIGRlc2M=)

[Sample report](https://s3.amazonaws.com/clickhouse-test-reports/0/ebf47dd7c2464cc5bc7bc1d49c6ad816ab9c617d/stateless_tests__aarch64_.html)

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Tagging as not for changelog since it was added  in https://github.com/ClickHouse/ClickHouse/pull/59545 which isn't released.

cc @davenger 

Closes https://github.com/ClickHouse/ClickHouse/issues/60190

### CODE REVIEW

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 1]
```diff
@@ -222,6 +222,7 @@ class SystemZooKeeperSource : public ISource
     ContextPtr context;
     ZooKeeperWithFaultInjection::Ptr zookeeper;
     bool started = false;
+    std::unordered_set<String> visited;
```
Comment (Algunenano) [inline src/Storages/System/StorageSystemZooKeeper.cpp:225]: This is the real fix. As generate() might be called more than once now, we need to keep this alive between calls to compare and not duplicate across blocks.

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 2]
```diff
@@ -169,13 +169,13 @@ class ZooKeeperSink : public SinkToStorage
...
-    Recurse, /// Fatch all nodes under this path, recursively
+    Exact, /// Fetch all nodes under this path
+    Prefix, /// Fetch all nodes starting with this prefix, recursively (multiple paths may match prefix)
+    Recurse, /// Fetch all nodes under this path, recursively
 };
 
-/// List of paths to be feched from zookeeper
-using Paths = std::deque<std::pair<String, ZkPathType>>;
+/// List of paths to be fetched from zookeeper
+using Paths = std::unordered_map<String, ZkPathType>;
```
Comment (Algunenano) [inline src/Storages/System/StorageSystemZooKeeper.cpp:178]: Use an unordered_map to deduplicate when possible before making requests to keeper

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 3]
```diff
@@ -443,8 +446,17 @@ static Paths extractPath(const ActionsDAG::NodeRawConstPtrs & filter_nodes, Cont
     for (const auto * node : filter_nodes)
         extractPathImpl(*node, res, context, allow_unrestricted);
 
+    auto node1 = res.find("/");
+    auto node2 = res.find("");
+    if ((node1 != res.end() && node1->second != ZkPathType::Exact) || (node2 != res.end() && node2->second != ZkPathType::Exact))
```
Comment (Algunenano) [inline src/Storages/System/StorageSystemZooKeeper.cpp:451]: With the old interpreter we have this:

```
EXPLAIN SYNTAX
SELECT
    name,
    path
FROM
(
    SELECT
        path,
        name
    FROM system.zookeeper
    WHERE path LIKE '/clickhouse/%'
    ORDER BY name ASC
)
WHERE path LIKE '%02221_system_zookeeper_unrestricted_like_test_00z68zd5/02221_system_zookeeper_unrestricted_like%'

Query id: 019e4581-a1fc-4428-8548-8749daa63dea

┌─explain───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ SELECT                                                                                                                                                    │
│     name,                                                                                                                                                 │
│     path                                                                                                                                                  │
│ FROM                                                                                                                                                      │
│ (                                                                                                                                                         │
│     SELECT                                                                                                                                                │
│         path,                                                                                                                                             │
│         name                                                                                                                                              │
│     FROM system.zookeeper                                                                                                                                 │
│     WHERE (path LIKE '/clickhouse/%') AND (path LIKE '%02221_system_zookeeper_unrestricted_like_test_00z68zd5/02221_system_zookeeper_unrestricted_like%') │
│     ORDER BY name ASC                                                                                                                                     │
│ )                                                                                                                                                         │
│ WHERE path LIKE '%02221_system_zookeeper_unrestricted_like_test_00z68zd5/02221_system_zookeeper_unrestricted_like%'                                       │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

13 rows in set. Elapsed: 0.015 sec.
```

As the condition has been pushed inside the inner query, the filter '%02221_system_zookeeper_unrestricted_like_test_00z68zd5/02221_system_zookeeper_unrestricted_like%' is already allowing everything. In fact it happens twice. So there is 3 filters, 2 visiting everything and one visiting `/clickhouse/%`. If we know we are going to visit everything we can forget about everything else.

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 8a161bdc0aa8b57c24dd43947081e80a38a16b70 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60429/8a161bdc0aa8b57c24dd43947081e80a38a16b70/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60429/8a161bdc0aa8b57c24dd43947081e80a38a16b70/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by davenger — APPROVED** [review]

**Pull request merged**